diff --git a/.gitignore b/.gitignore
index 49acc9f2e1..a6191d1d72 100644
--- a/.gitignore
+++ b/.gitignore
@@ -297,3 +297,16 @@ tmp-php.ini
 !/ext/fileinfo/magicdata.patch
 !/ext/pcre/pcre2lib/config.h
 !/win32/build/Makefile
+
+# ------------------------------------------------------------------------------
+# CMake build system files
+# ------------------------------------------------------------------------------
+/.ninja*
+/build.ninja
+/CMakeUserPresets.json
+/CTestTestfile.cmake
+/profile.json
+/Testing/
+cmake_install.cmake
+CMakeCache.txt
+CMakeFiles/
