diff --git a/cmake/CMakeLists.txt b/cmake/CMakeLists.txt
index 75ff5a0..40e753e 100644
--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -38,6 +38,10 @@ target_include_directories(
     ${PHP_SOURCE_DIR}
 )
 
+# Create a project wide INTERFACE library with static extensions.
+add_library(php_extensions INTERFACE)
+add_library(PHP::extensions ALIAS php_extensions)
+
 # Create a custom target for generating files (parsers, lexers, etc.) manually:
 #   cmake --build <dir> -t php_generate_files
 add_custom_target(php_generate_files)
diff --git a/cmake/TSRM/CMakeLists.txt b/cmake/TSRM/CMakeLists.txt
index 0f94593..8176666 100644
--- a/cmake/TSRM/CMakeLists.txt
+++ b/cmake/TSRM/CMakeLists.txt
@@ -1,6 +1,6 @@
 include(FeatureSummary)
 
-add_library(php_tsrm STATIC)
+add_library(php_tsrm OBJECT)
 add_library(PHP::TSRM ALIAS php_tsrm)
 
 target_sources(
diff --git a/cmake/Zend/CMakeLists.txt b/cmake/Zend/CMakeLists.txt
index f16a9ac..a6ff3d5 100644
--- a/cmake/Zend/CMakeLists.txt
+++ b/cmake/Zend/CMakeLists.txt
@@ -82,7 +82,7 @@ endif()
 # Add library.
 ################################################################################
 
-add_library(zend STATIC)
+add_library(zend OBJECT)
 add_library(Zend::Zend ALIAS zend)
 
 file(GLOB _zend_headers ${CMAKE_CURRENT_SOURCE_DIR}/*.h)
diff --git a/cmake/cmake/modules/PHP/Extensions.cmake b/cmake/cmake/modules/PHP/Extensions.cmake
index 4d3c83d..868856c 100644
--- a/cmake/cmake/modules/PHP/Extensions.cmake
+++ b/cmake/cmake/modules/PHP/Extensions.cmake
@@ -623,9 +623,9 @@ function(_php_extensions_post_configure directory)
     TARGETS php_${extension}
     ARCHIVE EXCLUDE_FROM_ALL
     RUNTIME
-      DESTINATION ${PHP_EXTENSION_DIR}
+      DESTINATION ${PHP_EXTENSION_DIR_REL}
     LIBRARY
-      DESTINATION ${PHP_EXTENSION_DIR}
+      DESTINATION ${PHP_EXTENSION_DIR_REL}
     FILE_SET HEADERS
       DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/ext/${extension}
   )
diff --git a/cmake/ext/spl/CMakeLists.txt b/cmake/ext/spl/CMakeLists.txt
index 1806260..63db8ca 100644
--- a/cmake/ext/spl/CMakeLists.txt
+++ b/cmake/ext/spl/CMakeLists.txt
@@ -1,4 +1,4 @@
-add_library(php_spl STATIC)
+add_library(php_spl OBJECT)
 
 target_sources(
   php_spl
diff --git a/cmake/ext/standard/CMakeLists.txt b/cmake/ext/standard/CMakeLists.txt
index 1d1ab13..2aed5da 100644
--- a/cmake/ext/standard/CMakeLists.txt
+++ b/cmake/ext/standard/CMakeLists.txt
@@ -26,7 +26,7 @@ add_feature_info(
 # Add library.
 ################################################################################
 
-add_library(php_standard STATIC)
+add_library(php_standard OBJECT)
 
 file(GLOB _php_standard_headers ${CMAKE_CURRENT_SOURCE_DIR}/*.h)
 # Exclude Windows header and specify it below separately.
@@ -122,8 +122,6 @@ target_link_libraries(
   php_standard
   PRIVATE
     $<$<PLATFORM_ID:Windows>:iphlpapi>
-    # TODO: Can this be fixed better?
-    PHP::main
 )
 
 ################################################################################
diff --git a/cmake/main/CMakeLists.txt b/cmake/main/CMakeLists.txt
index a51dad3..920c621 100644
--- a/cmake/main/CMakeLists.txt
+++ b/cmake/main/CMakeLists.txt
@@ -1,6 +1,6 @@
 include(FeatureSummary)
 
-add_library(php_main STATIC)
+add_library(php_main OBJECT)
 add_library(PHP::main ALIAS php_main)
 
 file(GLOB_RECURSE _php_main_headers ${CMAKE_CURRENT_SOURCE_DIR}/*.h)
@@ -13,7 +13,7 @@ target_sources(
     fopen_wrappers.c
     getopt.c
     internal_functions_cli.c
-    internal_functions.c
+    #internal_functions.c
     main.c
     network.c
     output.c
@@ -84,7 +84,7 @@ target_link_libraries(
   # Link publicly for PHP SAPIs.
   PUBLIC
     PHP::configuration
-    Zend::Zend
+    #Zend::Zend
 )
 
 if(TARGET PHP::win32)
@@ -187,7 +187,8 @@ foreach(extension IN LISTS extensions)
         POSITION_INDEPENDENT_CODE TRUE
     )
   else()
-    target_link_libraries(php_main PRIVATE PHP::${extension})
+    #target_link_libraries(php_main PRIVATE PHP::${extension})
+    target_link_libraries(php_extensions INTERFACE PHP::${extension})
   endif()
 endforeach()
 
diff --git a/cmake/pear/CMakeLists.txt b/cmake/pear/CMakeLists.txt
index 5ad2ab0..a080cd1 100644
--- a/cmake/pear/CMakeLists.txt
+++ b/cmake/pear/CMakeLists.txt
@@ -61,6 +61,8 @@ set(_php_pear_installer_url "https://pear.php.net/install-pear-nozlib.phar")
 
 # TODO: Recheck if _php_pear_dir needs to be manually created here.
 install(CODE "
+  message(STATUS \"$<TARGET_FILE:php_cli>\")
+  message(STATUS \"$<INSTALL_PREFIX>\")
   message(STATUS \"Installing PEAR environment: ${_php_pear_dir}\")
 
   if(EXISTS \"${CMAKE_CURRENT_SOURCE_DIR}/install-pear-nozlib.phar\"
diff --git a/cmake/sapi/cli/CMakeLists.txt b/cmake/sapi/cli/CMakeLists.txt
index bcd9f1b..2f942ac 100644
--- a/cmake/sapi/cli/CMakeLists.txt
+++ b/cmake/sapi/cli/CMakeLists.txt
@@ -72,8 +72,12 @@ target_compile_definitions(
 
 target_link_libraries(
   php_cli
-  PRIVATE
+  PUBLIC
+    php_standard
+    PHP::spl
     PHP::main
+    Zend::Zend
+    PHP::extensions
     $<$<PLATFORM_ID:Windows>:ws2_32;shell32>
 )
 
diff --git a/cmake/sapi/embed/CMakeLists.txt b/cmake/sapi/embed/CMakeLists.txt
index d2f07bc..da55d63 100644
--- a/cmake/sapi/embed/CMakeLists.txt
+++ b/cmake/sapi/embed/CMakeLists.txt
@@ -21,7 +21,23 @@ target_sources(
         php_embed.h
 )
 
-target_link_libraries(php_embed PRIVATE PHP::main)
+#[[
+target_link_libraries(
+  php_embed
+  PRIVATE
+    PHP::main
+    PHP::extensions
+)
+]]#
+target_link_libraries(
+  php_embed
+  PUBLIC
+    php_standard
+    PHP::spl
+    PHP::main
+    Zend::Zend
+    PHP::extensions
+)
 
 target_compile_definitions(php_embed PRIVATE ZEND_ENABLE_STATIC_TSRMLS_CACHE=1)
 
